freqtrade/freqai/data_drawer.py

```diff
@@ -0,0 +1,371 @@
+            mrv_df[f"{label}_std"] = dk.data["labels_std"][label]
+
+        if self.freqai_info.get("feature_parameters", {}).get("DI_threshold", 0) > 0:
+            mrv_df["DI_values"] = dk.DI_values
+
+        mrv_df["do_predict"] = do_preds
+
+    def append_model_predictions(self, pair: str, predictions, do_preds, dk, len_df) -> None:
+
+        # strat seems to feed us variable sized dataframes - and since we are trying to build our
+        # own return array in the same shape, we need to figure out how the size has changed
+        # and adapt our stored/returned info accordingly.
+        length_difference = len(self.model_return_values[pair]) - len_df
+        i = 0
+
+        if length_difference == 0:
+            i = 1
+        elif length_difference > 0:
+            i = length_difference + 1
+
+        df = self.model_return_values[pair] = self.model_return_values[pair].shift(-i)
+
+        if pair in self.historic_predictions:
+            hp_df = self.historic_predictions[pair]
+            # here are some pandas hula hoops to accommodate the possibility of a series
+            # or dataframe depending number of labels requested by user
+            nan_df = pd.DataFrame(np.nan, index=hp_df.index[-2:] + 2, columns=hp_df.columns)
+            hp_df = pd.concat([hp_df, nan_df], ignore_index=True, axis=0)
+            self.historic_predictions[pair] = hp_df[:-1]
+
+        for label in dk.label_list:
+            df[label].iloc[-1] = predictions[label].iloc[-1]
+            df[f"{label}_mean"].iloc[-1] = dk.data["labels_mean"][label]
+            df[f"{label}_std"].iloc[-1] = dk.data["labels_std"][label]
+        # df['prediction'].iloc[-1] = predictions[-1]
+        df["do_predict"].iloc[-1] = do_preds[-1]
+
+        if self.freqai_info.get("feature_parameters", {}).get("DI_threshold", 0) > 0:
+            df["DI_values"].iloc[-1] = dk.DI_values[-1]
+
+        # append the new predictions to persistent storage
+        if pair in self.historic_predictions:
+            self.historic_predictions[pair].iloc[-1] = df[label].iloc[-1]
+
+        if length_difference < 0:
+            prepend_df = pd.DataFrame(
+                np.zeros((abs(length_difference) - 1, len(df.columns))), columns=df.columns
+            )
+            df = pd.concat([prepend_df, df], axis=0)
+
+    def attach_return_values_to_return_dataframe(
+            self, pair: str, dataframe: DataFrame) -> DataFrame:
+        """
+        Attach the return values to the strat dataframe
+        :param dataframe: DataFrame = strategy dataframe
+        :return: DataFrame = strat dataframe with return values attached
+        """
+        df = self.model_return_values[pair]
+        to_keep = [col for col in dataframe.columns if not col.startswith("&")]
+        dataframe = pd.concat([dataframe[to_keep], df], axis=1)
+        return dataframe
+
+    def return_null_values_to_strategy(self, dataframe: DataFrame, dk) -> None:
```

Comment:
it's quite unfortunate that Datadrawer and datakitchen produce a cyclic import if we try to add typehints here.

Having the typehints here would greatly help mypy to identify potential issues - now they're all implicit "Any" - which is obviously a "can do everything and nothing" type.